app/src/main/cpp/native-lib.cpp

```diff
@@ -12,9 +38,208 @@ Java_com_example_cloudonix_MainActivity_stringFromJNI(
+             ifaddr_iterator != NULL; ifaddr_iterator = ifaddr_iterator->ifa_next) {
+            // Iterate IPv4 public addresses second
+            if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET)) {
+                struct sockaddr_in *in = (struct sockaddr_in *) ifaddr_iterator->ifa_addr;
+                getnameinfo(ifaddr_iterator->ifa_addr, sizeof(struct sockaddr_in6), addr,
+                            sizeof(addr), NULL, 0, NI_NUMERICHOST);
+//            sockaddr *tempaddr = ifaddr_iterator->ifa_addr;
+//            std::string address(tempaddr->sa_data);
+                std::string address(addr);
+                if (isIPv4Private(address)) {
+                    retVal = env->NewStringUTF(address.c_str());
+                }
+            }
+        }
+    }
+
+    // If we found nothing, keep looking...
+    if (retVal == nullptr) {
+        //Fallback case, return any IPv4 address we found. So we can return the first one.
+        for (ifaddr_iterator = ifaddr;
+             ifaddr_iterator != NULL; ifaddr_iterator = ifaddr_iterator->ifa_next) {
+            if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET)) {
+                struct sockaddr_in *in = (struct sockaddr_in *) ifaddr_iterator->ifa_addr;
+                getnameinfo(ifaddr_iterator->ifa_addr, sizeof(struct sockaddr_in6), addr,
+                            sizeof(addr), NULL, 0, NI_NUMERICHOST);
+
+                std::string address(addr);
+                retVal = env->NewStringUTF(address.c_str());
+            }
+        }
+    }
+
+    // Cleanup ifaddrs
+    freeifaddrs(ifaddr);
+
+    return retVal;
+}
+
+bool isIPv6GlobalUnicast(const std::string &addr) {
+    return addr.find("2000:") == 0;
+}
+
+bool isIPv4Private(const std::string &addr) {
+    return isIPv4PrivateCase1(addr)
+           || isIPv4PrivateCase2(addr)
+           || isIPv4PrivateCase3(addr)
+           || isIPv4PrivateLinkLocal(addr);
+}
+
+bool isIPv4PrivateCase1(const std::string &addr) {
+    // If the address starts with "10." it's on the 10.x.x.x. subnet, therefore it's private
+    return addr.find("10.") == 0;
+}
+
+bool isIPv4PrivateCase2(const std::string &addr) {
+    // Check for 172.16.0.0 - 172.31.255.255 case
+    // We should actually extract the 'octets' or whatever the name was...
+    // and then check if the first one is 172, and the second one is in 16-31 range
+
+    // Split the string by '.'
+    std::istringstream iss(addr);
+    std::string token;
+    std::vector<std::string> parts;
+    while (std::getline(iss, token, '.')) {
+        parts.push_back(token);
+    }
+
+    // Since we're here, we might as well do everything to validate the ipv4 IP address...
+    // which actually breaks the point of this method (checking if it's a private IP) so let's not.
+
+    // Check if the IP address has 4 parts
+//    if (parts.size() != 4) {
+//        return false;
+//    }
```

Comment:
get rid of this, it breaks the whole point of the method and completely changes it's semantics from "check if it's a private address" to "validate if it's a valid IP address".

Reply:
done

Reply:
resolving.